kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,230 @@
+        """Perform dimension reduction on x
+
+        Args:
+            x (tensor):
+            # reduce_mean (bool): Whether reduce the mean estimated from training data. Defaults to True.
+            return_vector (bool): Whether vectorize the tensor principal components. Defaults to False.
+            sort_by_var (bool): Whether sort features by variance in descending order. Defaults to True,
+                applies only when return_vector=True.
+            n_components (int): Number of components to keep. Applies only when return_vector=True.
+                Defaults to None.
+
+        Returns:
+            tensor
+                tensor principal components, shape (n_samples, P_1, P_2, ..., P_N) if return_vector==False.
+                If return_vector==True, features will be vectorized and sorted based on their explained
+                variance ratio, shape (n_samples, P_1 * P_2 * ... * P_N) if self.n_components is None,
+                and shape (n_samples, n_components) if self.n_component is not None.
+        """
+        _check_dim_shape(x, self.n_dim, self.shape_in)
+
+        # if reduce_mean:
+        #     x -= self.mean_
+
+        # tensor principal components
+        tensor_pc = multi_mode_dot(x, self.factors, modes=[m for m in range(1, self.n_dim)])
+
+        if return_vector:
+            tensor_pc = tl.unfold(tensor_pc, mode=0)
+            if sort_by_var:
+                tensor_pc = tensor_pc[:, self.idx_order]
+            if isinstance(n_components, int):
+                all_comp = int(torch.prod(torch.Tensor(self.shape_out)))
+                if n_components > all_comp:
+                    warnings.warn(
+                        "n_components can not exceed the multiplication of output "
+                        "tensor shapes, the maximum number of components will be applied."
+                    )
+                    n_components = all_comp
+                if not sort_by_var:
+                    tensor_pc = tensor_pc[:, self.idx_order]
+                tensor_pc = tensor_pc[:, :n_components]
+
+        return tensor_pc
+
+    def restore(self, x):
+        """Transform data in the shape of reduced dimension back to the original shape
+
+        Args:
+            x (tensor): Data to be restored, shape (n_samples, P_1, P_2, ..., P_N), where
+                P_1, P_2, ..., P_N are the reduced dimensions of of corresponding mode
+                (1, 2, ..., N), respectively. For tensors have been vectorized,
+                shape (n_samples, self.n_components) or (n_samples, P_1 * P_2 * ... * P_N).
+            # add_mean (bool): Whether add the mean estimated from the training set to the output.
+            #     Defaults to False.
+
+        Returns:
+            tensor:
+                Reconstructed tensor in original shape, shape (n_samples, I_1, I_2, ..., I_N).
+        """
+        if x.ndim <= 2:
+            if x.ndim == 1:
+                x = x.reshape((1, -1))
```

Comment:
If vector, then reshape into??? 
Are you doing vectorization here? You should do vectorization for x.dim == 2, right?
Maybe I am missing something here.

Reply:
If the input is a vector here, it will be reshaped into a matrix (one row, n columns). I need a 2D matrix (n_samples, P_1 * P_2 * ... * P_N, n_samples >=1) here to be folded as a tensor in shape (n_samples, P_1, P_2, ..., P_N).

Reply:
I see. Got it. I will help code viewer if some comments are added here.